network/agent/agent.spec.ts

```diff
@@ -89,3 +89,117 @@ test("don't use a proxy when the URL is in noProxy", () => {
   })
 })
 
+test('should return the correct client certificates', () => {
+  const agent = getAgent('https://foo.com/bar', {
+    clientCertificates: {
+      '//foo.com/': {
+        ca: 'ca',
+        cert: 'cert',
+        key: 'key',
+      },
+    },
+  })
+
+  expect(agent).toEqual({
+    ca: 'ca',
+    cert: 'cert',
+    key: 'key',
+    localAddress: undefined,
+    maxSockets: 50,
+    rejectUnauthorized: undefined,
+    timeout: 0,
+    __type: 'https',
+  })
+})
+
+test('should not return client certificates for a different host', () => {
+  const agent = getAgent('https://foo.com/bar', {
+    clientCertificates: {
+      '//bar.com/': {
+        ca: 'ca',
+        cert: 'cert',
+        key: 'key',
+      },
+    },
+  })
+
+  expect(agent).toEqual({
+    localAddress: undefined,
+    maxSockets: 50,
+    rejectUnauthorized: undefined,
+    timeout: 0,
+    __type: 'https',
+  })
+})
+
+test('scoped certificates override global certificates', () => {
+  const agent = getAgent('https://foo.com/bar', {
+    ca: 'global-ca',
+    key: 'global-key',
+    cert: 'global-cert',
+    clientCertificates: {
+      '//foo.com/': {
+        ca: 'scoped-ca',
+        cert: 'scoped-cert',
+        key: 'scoped-key',
+      },
+    },
+  })
+
+  expect(agent).toEqual({
+    ca: 'scoped-ca',
+    cert: 'scoped-cert',
+    key: 'scoped-key',
+    localAddress: undefined,
+    maxSockets: 50,
+    rejectUnauthorized: undefined,
+    timeout: 0,
+    __type: 'https',
+  })
+})
+
+test('select correct client certificates when host has a port', () => {
+  const agent = getAgent('https://foo.com:1234/bar', {
+    clientCertificates: {
+      '//foo.com:1234/': {
+        ca: 'ca',
+        cert: 'cert',
+        key: 'key',
+      },
+    },
+  })
+
+  expect(agent).toEqual({
+    ca: 'ca',
+    cert: 'cert',
+    key: 'key',
+    localAddress: undefined,
+    maxSockets: 50,
+    rejectUnauthorized: undefined,
+    timeout: 0,
+    __type: 'https',
+  })
+})
+
+test('select correct client certificates when host has a path', () => {
+  const agent = getAgent('https://foo.com/bar/baz', {
+    clientCertificates: {
+      '//foo.com/bar/': {
```

Comment:
This case should work too, should it?
```suggestion
  const agent = getAgent('https://foo.com/bar/baz', {
    clientCertificates: {
      '//foo.com/': {
```

Reply:
With this change the test fails, so maybe we need to add something for this, do we have something similar in the auth tokens? I think we fixed something like this a few weeks ago, I'm going to check it.

Reply:
https://github.com/pnpm/pnpm/commit/23039a6d60b48670c23774c8a2a0210509a4c138

Reply:
Not sure if it's the same issue though

Reply:
Ok, it's not the same, but we should use the same logic as the auth tokens
https://github.com/pnpm/pnpm/blob/23039a6d60b48670c23774c8a2a0210509a4c138/network/auth-header/src/index.ts#L27-L35

Reply:
Something like this fixes the error, what do you think?
```ts
function getClientCertificates(uri: string, opts: AgentOptions) {
  const parsedUri = new URL(uri);
  const host = nerfDart(uri);
  const path = parsedUri.pathname;
  const hostOnlyDomain = `//${parsedUri.hostname + (parsedUri.port ? `:${parsedUri.port}` : '')}/`

  if (host.endsWith(path)) {
    return opts.clientCertificates?.[host];
  }

  const fullPath = `${host}${path !== '/' ? path : ''}`;
  return opts.clientCertificates?.[fullPath] ?? opts.clientCertificates?.[host] ?? opts.clientCertificates?.[hostOnlyDomain];
}
```

Reply:
I think we should move the logic of reading URL-specific settings to a new package. The logic is in the `@pnpm/network.auth-header` package now.

Reply:
Should I create the new package inside this repo?